### PR TITLE
streaming downsample

### PR DESCRIPTION
**Pull request by smrutilal2**

### CODE REVIEW

File: core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java [hunk 1]
```diff
@@ -107,13 +107,36 @@ public void accumulate(final long[] values,
       }
     }
 
+    @Override
+    public void accumulate(double value, int idx) {
+      if (Double.isNaN(value)) {
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java]: We'll need the long version of the interface for the HBase stage but for this, do a `if (double_accumulator == null) <then allocate and copy from long_accumulator>` check in each `accumulate()` implementation.

File: core/src/main/java/net/opentsdb/data/types/numeric/aggregators/NumericArrayAggregator.java [hunk 2]
```diff
@@ -48,7 +48,14 @@ public void accumulate(final long[] values,
...
    */
   public void accumulate(final double[] values);
-  
+
+  /**
+   * Accumulate a double value at a specific index
+   * @param value
+   * @param index
+   */
+  public void accumulate(final double value, final int index);
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/data/types/numeric/aggregators/NumericArrayAggregator.java]: We'll need the long version for HBase but can leave it out for now.

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupBy.java [hunk 3]
```diff
@@ -44,9 +54,13 @@
  * @since 3.0
  */
 public class GroupBy extends AbstractQueryNode {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractQueryNode.class);
   
   /** The config for this group by node. */
   private final GroupByConfig config;
+
+  private DownsampleConfig downsampleConfig;
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/processor/groupby/GroupBy.java]: java doc it as `An optional downsample config for use by the GroupByArrayIterator to size it's array properly when running in parallel.`

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupBy.java [hunk 4]
```diff
@@ -63,7 +77,30 @@ public GroupBy(final QueryNodeFactory factory,
...
+                  TimeSeriesDataSourceConfig config = (TimeSeriesDataSourceConfig) timeSeriesDataSource.config();
+                  List<QueryNodeConfig> pushDownNodes = config.getPushDownNodes();
+                  for (QueryNodeConfig queryNodeConfig : pushDownNodes) {
+                      if(queryNodeConfig instanceof DownsampleConfig) {
+                        downsampleConfig = (DownsampleConfig) queryNodeConfig;
+                        break;
+                      }
+                  }
+                  break;
+                }
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/processor/groupby/GroupBy.java]: For the HBase case, also check to see if the node is an instance of `Downsample` and grab the config from there.

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java [hunk 5]
```diff
@@ -125,137 +120,131 @@ public GroupByNumericArrayIterator(final QueryNode node,
...
+    } else {
+      size = downsampleConfig.intervals();
     }
-    
-    logger.info("Sources size {}, Intervals found {}", sources.size(), intervals);
 
-    valuesCombiner = new NumericArrayAggregator[NUM_THREADS];
-    for (int i = 0; i < valuesCombiner.length; i++) {
-      valuesCombiner[i] = factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    logger.trace("Group size: " + size);
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java]: Yank or surround with `if (LOG.isTraceEnabled())`

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java [hunk 6]
```diff
@@ -125,137 +120,131 @@ public GroupByNumericArrayIterator(final QueryNode node,
...
+
+    aggregator =
+        factory.newAggregator(((GroupByConfig) node.config()).getInfectiousNan());
+    if (aggregator == null) {
+      throw new IllegalArgumentException(
+          "No aggregator found of type: " + ((GroupByConfig) node.config()).getAggregator());
     }
 
     // TODO: Need to check if it makes sense to make this threshold configurable
     if (sources.size() >= NUM_THREADS) {
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java]: Could probably do some factor of NUM_THREADS like `NUM_THREADS * 16`

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java [hunk 7]
```diff
@@ -125,137 +120,131 @@ public GroupByNumericArrayIterator(final QueryNode node,
...
-        }
+        accumulate(source, null);
       }
     }
-    
+  }
 
-    double e = ((double) (DateTime.nanoTime() - st)) / 1000_000;
-    logger.info("Groupby iterator took {}, timeseries count {}", e, sources.size());
+  private NumericArrayAggregator createAggregator(
```
Comment (manolama) [inline core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java]: Need's `final`s and we'll need a flag to determine if we should fill and set doubles or keep longs.

File: core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java (hunk 1 above)
Comment (smrutilal2) [reply core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java]: The long version of this api should be enough I guess

Comment (smrutilal2) [reply core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java]: I think initializing the array would be a prerequisite for this api.

File: core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java (hunk 6 above)
Comment (smrutilal2) [reply core/src/main/java/net/opentsdb/query/processor/groupby/GroupByNumericArrayIterator.java]: Remember, I commented on GSK's pr to make it configurable. This one is on you @manolama

File: core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java (hunk 1 above)
Comment (manolama) [reply core/src/main/java/net/opentsdb/data/types/numeric/aggregators/ArrayAverageFactory.java]: rgr

**Pull request merged**